content/blog/multi-single-tenant-github-app-installation/index.md

```diff
@@ -0,0 +1,62 @@
+
+# Background
+
+Roadie provides a SaaS version of the Backstage application that was open sourced by Spotify early last year. It is a single tenant application.
+
+In order to provide Backstage as a service we have settled on a multi single tenant architecture. This means that for each customer we provision a set of resources unique to the customers. We do have some shared resources between customers, however the main components of Backstage remain unique to each tenant. In the early stages we explored a purely multi tenanted application. However we felt that we would continue to fight against the tide of a single tenanted OpenSource Backstage project.
+
+![multi-single-tenant.png](multi-single-tenant.png)
+
+# Requirements
+
+Roadie provides a single GitHub App on the GitHub App marketplace called “Roadie Backstage”.
+
+The requirements we had around GitHub Apps was as follows:
+- The admin user on-boarding onto a newly created roadie backstage instance must be hinted to install the “Roadie Backstage” App
+- There must be absolute isolation between installations of the App. E.g., customer 1 must not be able to access the resources of customer 2 via Roadie.
+
+# Routing the Callback URL Redirect
+
+GitHub Apps provide a single “Callback URL” configuration option. When this is set, the user installing the backstage app is redirected to the configured Callback URL.
+
+Because there is one Callback URL we had to create a single service shared by all customers that could be used to handle the installation redirect. So we did just that. We created a lambda API service that we called the “redirect-handler”. It would accept a “state” parameter that contains the URL target of the redirect. The redirect handler would simply unpack the state parameter and return the user a redirect to the target, reproducing all of the available url params contained in the request.
+
+![redirect-handler.png](redirect-handler.png)
+
+As you might be able to decern by, the state variable is what the redirect-handler uses to decide where to send the user on the redirect. This state is sent to the user by the backstage application, when the user indicates that they want to install the GitHub App. The state variable is a JWT token signed by a private key owned by the Backstage instance. The redirect-handler is configured with the required trust store to trust JWT tokens signed by the customer tenants. The JWT tokens contain a payload item called "urn:roadie:rdt".
+
+![app-installation-flow.png](app-installation-flow.png)
+
+When the redirect-handler recieves the request, it is decoding the JWT token in the state variable, verifies it is signed by a trusted source, extracts the "urn:roadie:rdt" variable and sends the user back a redirect to there.
+
+Now that the customers backstage instance has received the redirect, Backstage must verify that it trusts this installation.
+
+# Trusting the Installation ID returned via a redirect
+
```

Comment:
I was thinking that we could mention that we worked with env0 on coming to this solution.

Reply:
That's a very good idea.

Reply:
@dtuite, what do you think of this? I mean they did come up with the suggestion for us in the end. Would it be appropriate to name drop?

Reply:
I'm happy to mention them but I think you'd have to ask them first if they want to be mentioned.

Reply:
OK. I think I wont.